examples/step-19/step-19.cc

```diff
@@ -468,52 +467,16 @@ namespace Step19
         //   <a href="#extensions">"possibilities for extensions" section</a>
         //   below, and use a better approach in step-70, for example.
         if (particle_handler.n_particles_in_cell(cell) > 0)
-          {
-            std::vector<Point<dim>> particle_reference_locations;
-
-            const typename Particles::ParticleHandler<
-              dim>::particle_iterator_range particles_in_cell =
-              particle_handler.particles_in_cell(cell);
-
-            const unsigned int n_particles_in_cell =
-              particle_handler.n_particles_in_cell(cell);
-
-            particle_reference_locations.resize(n_particles_in_cell);
-
+          for (const auto &particle : particle_handler.particles_in_cell(cell))
             {
-              typename Particles::ParticleHandler<dim>::particle_iterator
-                particle = particles_in_cell.begin();
-              for (unsigned int particle_index = 0;
-                   particle != particles_in_cell.end();
-                   ++particle, ++particle_index)
-                particle_reference_locations[particle_index] =
-                  particle->get_reference_location();
-            }
-
-            // Now that we know where the particles on the current cell are
-            // located with regard to the reference cell's coordinate system, we
-            // can create a Quadrature object with these locations and then an
-            // FEValues object that we will use to evaluate the shape functions
-            // at these locations. The contribution to the right hand side
-            // vector then immediately follows from the formula shown above.
-            // Note again the absence of the call to FEValues::JxW that would
-            // have to be present if we were evaluating an integral.
-            const Quadrature<dim> quadrature_formula(
-              particle_reference_locations);
-            FEValues<dim> fe_value(mapping,
-                                   fe,
-                                   quadrature_formula,
-                                   update_values);
-
-            for (const unsigned int q_index :
-                 fe_values.quadrature_point_indices())
+              const Point<dim> reference_location =
+                particle.get_reference_location();
               for (const unsigned int i : fe_values.dof_indices())
                 cell_rhs(i) +=
-                  (fe_values.shape_value(i, q_index) *   // phi_i(x_p)
-                   (-Constants::electrons_per_particle * // N
-                    Constants::electron_charge));        // e
-          }
-
+                  (fe.shape_value(i, reference_location) * // phi_i(x_p)
+                   (-Constants::electrons_per_particle *   // N
+                    Constants::electron_charge));          // e
+            }
```

Comment:
I did not like the old code at all because the constructor of `FEValues` is so horrendously expensive. Here it is really simple (given that we use a `FiniteElement` whose values do not depend on the position), we should simply use the *virtual* `FiniteElement::shape_value()` function, much faster than the alternative. Of course, the really interesting place would be the one below where we need the gradients which depend on the mapping. Given that we have many particle codes, we should think of a better way to compute mapped gradients in those position. The `FEValues` interface is inappropriate in my opinion, even though it is quite powerfule.